File: app/src/main/java/com/pleon/buyt/database/AppDatabase.java
package com.pleon.buyt.database;

import android.content.Context;

import com.pleon.buyt.database.dao.ItemDao;
import com.pleon.buyt.database.dao.PurchaseDao;
import com.pleon.buyt.database.dao.StoreDao;
import com.pleon.buyt.model.Item;
import com.pleon.buyt.model.Purchase;
import com.pleon.buyt.model.Store;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.room.TypeConverters;

// Usually, you only need one instance of the Room database for the whole app
@Database(entities = {Item.class, Store.class, Purchase.class}, version = 1)
@TypeConverters({DateConverter.class, QuantityUnitConverter.class})
public abstract class AppDatabase extends RoomDatabase {

    private static final String DATABASE_NAME = "buyt-database.db";

    private static volatile AppDatabase instance;

    public static AppDatabase getDatabase(Context context) {
        if (instance == null) { // double-checked locking
            synchronized (AppDatabase.class) {
                if (instance == null) {
                    context = context.getApplicationContext();
                    instance = Room.databaseBuilder(context, AppDatabase.class, DATABASE_NAME)
//                            .fallbackToDestructiveMigration()
                            .build();
                }
            }
        }
        return instance;
    }

    // TODO: call this method when appropriate
    public static void destroyInstance() {
        instance = null;
    }

    public abstract ItemDao itemDao();

    public abstract PurchaseDao purchaseDao();

    public abstract StoreDao storeDao();
}

File: app/src/main/kotlin/com/pleon/buyt/database/AppDatabase.kt
package com.pleon.buyt.database

import androidx.room.Database
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import com.pleon.buyt.database.converter.CategoryConverter
import com.pleon.buyt.database.converter.DateConverter
import com.pleon.buyt.database.converter.QuantityUnitConverter
import com.pleon.buyt.database.dao.*
import com.pleon.buyt.model.Item
import com.pleon.buyt.model.Purchase
import com.pleon.buyt.model.Store
import com.pleon.buyt.model.Subscription

const val DB_NAME = "buyt-database.db"
const val DB_VERSION = 1

@Database(
    entities = [
        Item::class,
        Store::class,
        Purchase::class,
        Subscription::class
    ],
    version = DB_VERSION,
    // Export database schema info to have the generated SQL code and to keep a history of it
    // The location of the exported files can be set in the app build script
    exportSchema = true
)
@TypeConverters(
    DateConverter::class,
    CategoryConverter::class,
    QuantityUnitConverter::class
)
abstract class AppDatabase : RoomDatabase() {

    abstract fun itemDao(): ItemDao

    abstract fun storeDao(): StoreDao

    abstract fun purchaseDao(): PurchaseDao

    abstract fun databaseDao(): DatabaseDao

    abstract fun subscriptionDao(): SubscriptionDao
}
